### PR TITLE
Fix nunique for `MultiIndex`, `DataFrame`, and all NA case with `dropna=False`

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
Fixes 3 bugs with `nunique`

* `MultiIndex.nunique` returning a `dict` instead of an `int`
* `.nunique(dropna=False)` with all `NA`s returning 0 instead of 1
* `DataFrame.nunique` preserving column class and type in the resulting `Series.index`

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/tests/test_series.py [hunk 1]
```diff
@@ -2848,3 +2848,13 @@ def test_nans_to_nulls_noop_copies_column(value):
     ser1 = cudf.Series([value])
     ser2 = ser1.nans_to_nulls()
     assert ser1._column is not ser2._column
+
+
+@pytest.mark.parametrize("dropna", [False, True])
+def test_nunique_all_null(dropna):
+    data = [np.nan, np.nan]
+    pd_ser = pd.Series(data)
+    cudf_ser = cudf.Series(data, nan_as_null=True)
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_series.py:2857]: I don't think we are actually testing `nan` case here or do you want to test only `NA` cases?

File: python/cudf/cudf/tests/test_multiindex.py [hunk 2]
```diff
@@ -2162,3 +2162,14 @@ def test_multi_index_contains_hashable():
         lfunc_args_and_kwargs=((),),
         rfunc_args_and_kwargs=((),),
     )
+
+
+@pytest.mark.parametrize("array", [[1, 2], [1, np.nan], [np.nan] * 2])
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_multiindex.py:2167]: Same concern as below

**Review by galipremsagar — CHANGES_REQUESTED** [review]
We will have to investigate 1+% decrease in pass rate too..

File: python/cudf/cudf/tests/test_series.py (hunk 1 above)
Comment (mroeschke) [reply python/cudf/cudf/tests/test_series.py:2857]: I was hoping to just test the NA case here i.e. where `dropna` would impact the nunique count

**Comment by mroeschke** [comment]
> We will have to investigate 1+% decrease in pass rate too..

As of 72b7ac3ba2691c7427aebfe133e0792bad3678f2, it looks like the diff job is reporting a 0% change. https://github.com/rapidsai/cudf/actions/runs/9486926113

(Looks like there's still some noise in test failures passing between commits e.g. https://github.com/rapidsai/cudf/actions/runs/9473554707 vs https://github.com/rapidsai/cudf/actions/runs/9486926113)

File: python/cudf/cudf/tests/test_dataframe.py [hunk 3]
```diff
@@ -9966,6 +9966,20 @@ def test_dataframe_nunique(data):
...
+    [
+        pd.RangeIndex(2, name="foo"),
+        pd.MultiIndex.from_arrays([[1, 2], [2, 3]], names=["foo", 1]),
+        pd.Index([3, 5], dtype=np.int8, name="foo"),
+    ],
+)
+def test_nunique_preserve_column_in_index(columns):
+    df = cudf.DataFrame([[1, 2]], columns=columns)
+    result = df.nunique().index.to_pandas()
+    pd.testing.assert_index_equal(result, columns, exact=True)
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_dataframe.py:9980]:
```suggestion
    df = cudf.DataFrame([[1, 2]], columns=columns)
    result = df.nunique().index.to_pandas()
    assert_eq(result, columns, exact=True)
```

File: python/cudf/cudf/tests/test_series.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_series.py:2860]: ah, okay. Can we use `None` then instead of `np.nan`?

Comment (mroeschke) [reply python/cudf/cudf/tests/test_series.py]: Sure thing. Changed to use None

**Review by galipremsagar — APPROVED** [review]
python approval. I think we need two cpp approvals for the libcudf changes.

File: cpp/src/stream_compaction/distinct_count.cu [hunk 4]
```diff
@@ -187,7 +187,15 @@ cudf::size_type distinct_count(column_view const& input,
...
 {
-  if (0 == input.size() or input.null_count() == input.size()) { return 0; }
+  if (0 == input.size()) { return 0; }
+
+  if (input.null_count() == input.size()) {
+    if (null_handling == null_policy::INCLUDE) {
+      return 1;
+    } else {
+      return 0;
+    }
```
Comment (davidwendt) [inline cpp/src/stream_compaction/distinct_count.cu:197]:
```suggestion
  if (input.null_count() == input.size()) { return static_cast<size_type>(null_handling == null_policy::INCLUDE); }
```

**Review by davidwendt — APPROVED** [review]
Minor suggestion.

File: python/cudf/cudf/tests/test_multiindex.py [hunk 5]
```diff
@@ -2162,3 +2162,14 @@ def test_multi_index_contains_hashable():
         lfunc_args_and_kwargs=((),),
         rfunc_args_and_kwargs=((),),
     )
+
+
+@pytest.mark.parametrize("array", [[1, 2], [1, None], [None] * 2])
```
Comment (bdice) [inline python/cudf/cudf/tests/test_multiindex.py:2167]: Slightly less cognitive overhead.
```suggestion
@pytest.mark.parametrize("array", [[1, 2], [1, None], [None, None]])
```

File: python/cudf/cudf/core/multiindex.py [hunk 6]
```diff
@@ -1746,6 +1746,14 @@ def fillna(self, value):
     def unique(self):
         return self.drop_duplicates(keep="first")
 
+    @_cudf_nvtx_annotate
+    def nunique(self, dropna: bool = True) -> int:
+        if dropna:
+            mi = self.dropna(how="all")
+        else:
+            mi = self
```
Comment (bdice) [inline python/cudf/cudf/core/multiindex.py:1754]:
```suggestion
        mi = self.dropna(how="all") if dropna else self
```

**Review by bdice — APPROVED** [review]
A few minor suggestions, generally looks good.

**Review by shrshi — APPROVED** [review]
Looks good to me!

File: cpp/src/stream_compaction/distinct_count.cu [hunk 7]
```diff
@@ -187,7 +187,10 @@ cudf::size_type distinct_count(column_view const& input,
                                nan_policy nan_handling,
                                rmm::cuda_stream_view stream)
 {
-  if (0 == input.size() or input.null_count() == input.size()) { return 0; }
+  if (0 == input.size()) { return 0; }
+
+  if (input.null_count() == input.size()) { return static_cast<size_type>(null_handling == null_policy::INCLUDE); }
+  }
```
Comment (mroeschke) [inline cpp/src/stream_compaction/distinct_count.cu:193]:
```suggestion
```

**Pull request merged**